### ISSUE TITLE
Overfitting issues while using pre-trained weights

### ISSUE THREAD

**Issue by rees8** [ultralytics/ultralytics#12794]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm training a custom dataset using pre-trained weights for YOLOv8. 
The lines of code I used:

from ultralytics import YOLO
model = YOLO('yolov8n.pt')  
results = model.train(data='/kaggle/input/isl-images/ISL_DATASET_DRIVE/data.yaml', epochs=100, imgsz=640, device=0)

I have implemented the trained model into an Android application and it seems to recognize the images it's been trained on but not any new images. I believe this is a case of overfitting. I am really new to YOLO and I'm not sure how to reduce overfitting while using pre-trained weights since the parameters are not being explicitly mentioned.

I would like to know how do I approach this problem and if possible, provide a sample code snippet that I can experiment with.

**Comment by glenn-jocher** [comment]
Hello! It looks like you're facing overfitting with your custom-trained YOLOv8 model. Overfitting often occurs when the model learns the training data too well, including its noise and details, at the expense of its ability to generalize to new data.

Here are a few strategies you might consider to reduce overfitting:

1. **Data Augmentation**: Enrich your training dataset by transforming the training images (e.g., rotations, flipping, scaling). This helps the model generalize better.
2. **Regularization Techniques**: Try using techniques like dropout, or adjusting the learning rate.
3. **Early Stopping**: Monitor the model's performance on a validation set and stop training before it begins to over-fit.

Here's a code snippet that incorporates data augmentation, using the configuration file to specify transformations:

```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.pt')

# Specify augmentation policies in your dataset YAML file
# Here assuming you have augmentation options set up in your data.yaml
results = model.train(data='/kaggle/input/isl-images/ISL_DATASET_DRIVE/data.yaml', 
                      epochs=100, imgsz=640, device=0)
```

Ensure your `data.yaml` includes augmentation settings. Check the Ultralytics documentation for more on how to set this up.

These steps should help mitigate overfitting, giving your model a better chance to generalize to new images not seen during training. Give it a try and see how it improves your application's performance. Good luck! 🚀

**Comment by rees8** [comment]
Thank you, will try it out. Just as a double confirmation if this is indeed a case of overfitting or if there are some other issues, I am also attaching the performance metrics of the trained model: 

![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/88110289/d3f609ca-07f7-4e27-a5c5-d52a6535fe1d)
![F1_curve](https://github.com/ultralytics/ultralytics/assets/88110289/08ffd621-829b-4bdd-a767-2caed6e1eee6)
![P_curve](https://github.com/ultralytics/ultralytics/assets/88110289/185e58c8-ec87-48f6-ba01-29b91c87c28f)
![PR_curve](https://github.com/ultralytics/ultralytics/assets/88110289/0de2ed8d-6d2c-49bf-8c08-c80f7cd3744f)
![R_curve](https://github.com/ultralytics/ultralytics/assets/88110289/c988c016-a218-4583-9a76-dd36f811c78c)
![results](https://github.com/ultralytics/ultralytics/assets/88110289/c14fe45a-0bb3-470a-94b3-273733b2e3ea)

**Comment by glenn-jocher** [comment]
Thanks for sharing the performance metrics! 📊 Based on these charts, indeed it appears there could be some overfitting, as suggested by the narrow generalization gap on the PR curve and likely some biases in classification (as hinted by the confusion matrix). You might also want to explore if there's any class imbalance or data inconsistency which could affect the model.

Implementing the earlier suggested strategies might help in improving the generalization. Additionally, experimenting with different varieties of data augmentation could also prove beneficial. If adjustments in training don't fully resolve the issue, consider reviewing the dataset's diversity and representation.

Keep experimenting, and feel free to reach out if you encounter further issues. Good luck! 🚀

**Comment by rees8** [comment]
I came across a blog where it mentioned that Roboflow offers augmentation options while creating the dataset, so I combined my custom dataset and one from Roboflow to have variations and then applied augmentations like shear, saturation, crop, etc. My dataset size was previously around 3388 images in train and 569 images in valid. After applying augmentations, there are 12438 images in train and 1036 in valid. I began re-training the model for 150 epochs but stopped at 130 as the P and R values reached a plateau for the last 4 epochs. My model is starting to recognize the objects in real-time now. I am attaching the new performance metrics below. What I want to know is that is this model good enough to be included in my results section (this is my capstone project) or would you suggest and further changes? I am not quite aware of what the ideal metrics should be so I'm not able to make the decision. This model's metrics seem to be too good to be true, but again, it's performing better than before.
(Sorry for bombarding you with such seemingly basic questions)

![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/88110289/a977094c-b111-45b3-8658-c24abd1adec0)
![F1_curve](https://github.com/ultralytics/ultralytics/assets/88110289/87bf3cda-e955-4ebf-9e6d-ab091ef6f172)
![P_curve](https://github.com/ultralytics/ultralytics/assets/88110289/1f3e7ae5-16cf-416a-988c-1e08af2a8156)
![PR_curve](https://github.com/ultralytics/ultralytics/assets/88110289/fb37a0c3-de8b-4c6d-adb5-8fecff14d8fb)
![R_curve](https://github.com/ultralytics/ultralytics/assets/88110289/cc9f8d70-1a4d-4c75-8a13-0288d91496f9)

**Comment by glenn-jocher** [comment]
@rees8 hello! It's great to hear about your progress and the improvements in your model's performance after applying augmentations. 🚀 From the metrics and the confusion matrix you've shared, it seems like your model is performing quite well, especially considering the significant increase in training data and the diversity introduced by augmentations.

For your capstone project, if the model is performing well in real-time and the metrics are stable (as indicated by the plateau in P and R values), it sounds like a solid candidate to include in your results section. The key here is the consistency of performance across different scenarios and the robustness your augmentations have introduced.

If you're looking for further validation, you might consider:
- Conducting a few more tests in different real-world conditions to ensure the model's robustness.
- Comparing these results with a baseline model or previous iterations to highlight the improvements.

Your approach and the results look promising! Don't hesitate to include these findings in your project, and good luck with your capstone presentation! 🌟

**Comment by rees8** [comment]
Thanks a lot for your assistance!

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions or need further assistance as you continue working with YOLOv8, feel free to reach out. Happy coding! 😊
